core/src/autogluon/core/models/abstract/abstract_model.py

```diff
@@ -573,7 +573,15 @@ def _preprocess_fit_resources(self, silent=False, total_resources=None, parallel
                 )
                 minimum_model_num_cpus = minimum_model_resources.get('num_cpus', 1)
                 enforced_num_cpus = max(minimum_model_num_cpus, enforced_num_cpus - 2)  # leave some cpu resources for process running by cluster nodes
-                kwargs["num_cpus"] = enforced_num_cpus
+            max_resources = self._get_maximum_resources()
+            max_num_cpus = max_resources.get("num_cpus", None)
+            max_num_gpus = max_resources.get("num_gpus", None)
+            if max_num_cpus is not None:
+                enforced_num_cpus = min(max_num_cpus, enforced_num_cpus)
+            if max_num_gpus is not None:
+                enforced_num_gpus = min(max_num_gpus, enforced_num_gpus)
+            kwargs["num_cpus"] = enforced_num_cpus
+            kwargs["num_gpus"] = enforced_num_gpus
```

Comment:
What if max_num_gpus=0, but enforced_num_gpus=1? Currently it appears like we have `minimum_model_resources` that is passed `enforced_num_gpus > 0`. Should this instead be passed that value after it is updated by `max_num_gpus`?

Reply:
You right. This does become tricky though cause we are trying to reserve some cpus in the distributed scenario. We do so first and check if that's still above the max. If we reverse it, we'll be taking 2 more from the max. I guess this still sort of makes sense? But to avoid it, we need to check the total cpu available for the system and check if we are already using 2 less than it

Reply:
We can't take 2 more from the max, imagine if `max_cpus=1`, this would crash.

Reply:
I'm saying to have 

```
if max_num_gpus is not None:
                enforced_num_gpus = min(max_num_gpus, enforced_num_gpus)
```
Prior to the `if DistributedContext` logic.

The `max_num_cpus` logic can be after the `if DistributedContext` logic.

Reply:
sure